<a href="https://colab.research.google.com/github/forhadmia231/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/forhadmia231/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [28]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
!git clone https://github.com/forhadmia231/flyrank-ml-internship.git
%cd flyrank-ml-internship
!find . -type f

Cloning into 'flyrank-ml-internship'...
remote: Enumerating objects: 121, done.
remote: Counting objects: 100% (121/121), done.
remote: Compressing objects: 100% (77/77), done.
remote: Total 121 (delta 35), reused 96 (delta 28), pack-reused 0 (from 0)
Receiving objects: 100% (121/121), 1.82 MiB | 9.88 MiB/s, done.
Resolving deltas: 100% (35/35), done.
/content/flyrank-ml-internship/flyrank-ml-internship
./notebooks/02_your_first_readable_model.ipynb
./notebooks/03_working_with_the_full_release.ipynb
./notebooks/01_first_look_and_discovery.ipynb
./docs/ml-core-foundation-framework.md
./docs/ml-intern-dataset-and-lane-guide.md
./docs/intern-free-tooling-guide.md
./docs/flyrank-seo-research-march-2026.pdf
./docs/data-dictionary.md
./requirements.txt
./DATA_USE.md
./.git/refs/heads/main
./.git/refs/remotes/origin/HEAD
./.git/config
./.git/objects/pack/pack-eeeac607a1e8ad3f53a8298850865a8d1be4264e.idx
./.git/objects/pack/pack-eeeac607a1e8ad3f53a8298850865a8d1be4264e.pack
./.git/HEAD
./.git/

In [29]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

df.head()

,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [30]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

df.info()

Rows: 30000
Columns: 44
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 44 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   content_id              30000 non-null  object 
 1   client_id               30000 non-null  object 
 2   search_volume           27532 non-null  float64
 3   competition             27532 non-null  float64
 4   competition_level       27390 non-null  object 
 5   cpc                     27532 non-null  float64
 6   content_type            30000 non-null  object 
 7   main_intent             27626 non-null  object 
 8   word_count              22301 non-null  float64
 9   char_count              22301 non-null  float64
 10  provider_used           8562 non-null   object 
 11  model_used              24267 non-null  object 
 12  impressions_90d         30000 non-null  int64  
 13  clicks_90d              30000 non-null  int64  
 14  pageviews_90d 

In [31]:
print(df.columns.tolist())

['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


# 1. My Lane

## Selected Lane
Learning to Rank (Search Ranking Optimization)

## Why I chose this lane

I chose the Learning to Rank lane because search ranking directly influences content visibility, user engagement, and website traffic. By analyzing search performance data, I want to understand which content should be prioritized to improve search rankings and user experience. This project also aligns with my interest in machine learning and search optimization.

# 2. The Question

## Research Question

Can machine learning identify which content pages should be prioritized for refresh in order to improve search ranking and user engagement?

## Decision

Which content pages should be refreshed first?

## Action

SEO teams can prioritize updating pages with high search potential but declining performance, helping improve rankings and organic traffic.

## Cost of a Wrong Recommendation

If the recommendation is wrong, valuable content may be ignored while low-impact pages receive unnecessary updates. This wastes time, resources, and may reduce search visibility.

## Why Machine Learning?

Search performance depends on multiple interacting factors such as search volume, CTR, average position, engagement rate, content age, and trend direction. Machine learning can learn these patterns and provide data-driven recommendations for prioritizing content refresh.

In [32]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

print("\nAverage CTR:", round(df["ctr"].mean(),2))
print("Average Position:", round(df["avg_position"].mean(),2))
print("Average Search Volume:", round(df["search_volume"].mean(),2))

Rows: 30000
Columns: 44

Average CTR: 0.51
Average Position: 16.34
Average Search Volume: 158.88


In [33]:
df[["ctr","avg_position","search_volume","engagement_rate"]].describe()

,ctr,avg_position,search_volume,engagement_rate
count,30000.000000,30000.00000,27532.000000,30000.000000
mean,0.510733,16.34238,158.882391,2.534520
std,3.279162,15.21679,1518.270825,8.310096
min,0.000000,0.00000,0.000000,0.000000
25%,0.000000,6.20000,0.000000,0.000000
50%,0.070000,10.80000,10.000000,0.000000
75%,0.290000,22.30000,20.000000,1.350000
max,100.000000,245.00000,74000.000000,100.000000


# 3. Quick Look at the Data

The starter dataset contains **30,000** content records and **44** features.

The average click-through rate (CTR) is **0.511 (51.07%)**.

The average search position is **16.34**.

The average search volume is **158.88**.

These values show that search performance varies across different content pages. This variation suggests that machine learning could help identify which pages should be prioritized for content refresh and ranking improvements.

# 4. Careful Words

This project aims to identify patterns that can support content ranking decisions.

The model cannot guarantee that updating a page will improve rankings.

The recommendations are based on historical data and should be used as decision support rather than absolute truth.

This project identifies correlations, not causal relationships.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


# 5. Self Check

- [x] I selected a lane.
- [x] I defined a research question.
- [x] I described the decision.
- [x] I explained the action.
- [x] I described the cost of a wrong recommendation.
- [x] I included real numbers from the dataset.
- [x] I avoided making unsupported claims.

In [34]:
print(df.shape)
print(df["ctr"].mean())
print(df["avg_position"].mean())
print(df["search_volume"].mean())

(30000, 44)
0.5107333333333334
16.342380000000002
158.88239139909922


In [35]:
# Dataset Overview

print("Dataset Shape:", df.shape)

print("\nAverage CTR:", round(df["ctr"].mean(),3))
print("Average Position:", round(df["avg_position"].mean(),2))
print("Average Search Volume:", round(df["search_volume"].mean(),2))

print("\nContent Types:")
print(df["content_type"].value_counts())

print("\nIntent Distribution:")
print(df["main_intent"].value_counts())

Dataset Shape: (30000, 44)

Average CTR: 0.511
Average Position: 16.34
Average Search Volume: 158.88

Content Types:
content_type
keyword article       27207
feedly article         2096
comparison article      697
Name: count, dtype: int64

Intent Distribution:
main_intent
informational    17235
transactional     5733
commercial        4612
navigational        46
Name: count, dtype: int64
